src/proxy_handler.cc

```diff
@@ -0,0 +1,180 @@
+#include "proxy_handler.h"
+
+#include <string>
+
+RequestHandler::Status ProxyHandler::Init(const std::string& uri_prefix,
+    const NginxConfig& config)
+{
+  uri_prefix_ = uri_prefix;
+
+  int server_location = 0;
+  if (config.statements_[0]->tokens_[0] == "remote_port") {
+    server_location = 1;
+  }
+  remote_host_ =
+    config.statements_[server_location]->tokens_[1];
+  remote_port_ =
+    config.statements_[1-server_location]->tokens_[1];
+
+  return RequestHandler::OK;
+}
+
+RequestHandler::Status ProxyHandler::HandleRequest(const Request& request,
+    Response* response)
+{
+  if (remote_host_.size() == 0 || remote_port_.size() == 0) {
+    std::cerr << "ProxyHandler::HandleRequest: remote_host_ or remote_port_ ";
+    std::cerr << "are empty. Cannot serve request." << std::endl;
+    return RequestHandler::Error;
+  }
+
+  Request proxied_req(request);
+  proxied_req.set_uri(request.uri().substr(uri_prefix_.size()));
+  if (proxied_req.uri().size() == 0) {
+    proxied_req.set_uri("/");
+  }
+  // make sure keep alive is off
+  proxied_req.remove_header("Connection");
+  proxied_req.add_header("Connection", "Close");
+  // update host
+  proxied_req.remove_header("Host");
+  proxied_req.add_header("Host", remote_host_ + ":" + remote_port_);
+
+  client_ = new (std::nothrow) SyncClient();
```

Comment:
Actually it's deleted in the destructor defined in the header file. Still I changed it to unique_ptr since we're using C++11 after all.